### PR TITLE
pkg/gtid: remove gtid.Set interface

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close #4287 

### What is changed and how it works?

compared with mysql.GTIDSet, gtid.Set has two more methods. 
- `Replace` is no longer used
- `Truncate` is used but does not take effects, see https://github.com/pingcap/dm/pull/335#issuecomment-1106069436

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has interface methods change

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
 `None`.
```

### CODE REVIEW

**Comment by lance6716** [comment]
I'll write a better copy-on-write API for GTID set soon

File: dm/syncer/checkpoint.go [hunk 1]
```diff
@@ -71,7 +72,15 @@ type tablePoint struct {
 }
 
 func (b *tablePoint) String() string {
-	return fmt.Sprintf("location(%v), tableInfo(ID: %d, Name:%s, ColNum: %d, IdxNum: %d, PKIsHandle: %t)", b.location, b.ti.ID, b.ti.Name, len(b.ti.Columns), len(b.ti.Indices), b.ti.PKIsHandle)
+	if b == nil {
+		return ""
+	}
+	var buf strings.Builder
+	buf.WriteString(fmt.Sprintf("location(%s)", b.location))
```
Comment (lichunzhu) [inline dm/syncer/checkpoint.go:79]:
```suggestion
	fmt.Fprintf(&buf, "location(%s)", b.location)
```

File: dm/syncer/checkpoint.go [hunk 2]
```diff
@@ -71,7 +72,15 @@ type tablePoint struct {
...
+	if b.ti != nil {
+		buf.WriteString(fmt.Sprintf(", tableInfo(ID: %d, Name:%s, ColNum: %d, IdxNum: %d, PKIsHandle: %t)", b.ti.ID, b.ti.Name, len(b.ti.Columns), len(b.ti.Indices), b.ti.PKIsHandle))
```
Comment (lichunzhu) [inline dm/syncer/checkpoint.go:81]: ditto

File: dm/syncer/checkpoint.go [hunk 3]
```diff
@@ -183,7 +192,7 @@ func (b *binlogPoint) String() string {
 	b.RLock()
 	defer b.RUnlock()
 
-	return fmt.Sprintf("%v(flushed %v)", b.savedPoint, b.flushedPoint)
+	return fmt.Sprintf("%s(flushed %s)", b.savedPoint.String(), b.flushedPoint.String())
```
Comment (lichunzhu) [inline dm/syncer/checkpoint.go:195]: ditto

**Review by lichunzhu — APPROVED** [review]
LGTM

File: dm/pkg/binlog/position.go [hunk 4]
```diff
@@ -398,37 +411,23 @@ func (l *Location) ResetSuffix() {
 	l.Suffix = 0
 }
 
-// SetGTID set new gtid for location
-// Use this func instead of GITSet.Set to avoid change other location.
+// SetGTID set new gtid for location.
+// TODO: don't change old Location and return a new one to copy-on-write.
```
Comment (Ehco1996) [inline dm/pkg/binlog/position.go:415]: >  don't change old Location and return a new one to copy-on-write

I can't remember why we need to do this, can you please give me some tips? is this about the func`location.CloneWithFlavor` ?

**Review by Ehco1996 — COMMENTED** [review]
lgtm

File: dm/pkg/binlog/position.go (hunk 4 above)
Comment (lance6716) [reply dm/pkg/binlog/position.go:415]: it's a new request. mainly about reducing the possibility that when two variables hold the same reference to a location, changing one of them will accidentally change the other one: SetGTID should not change the input location. I think new API is less error-prone

this problem has another solutions, for example, we use location struct not location pointer, so if a location is assigned to another variables they are different copies. we can choose one solution in following pr.

**Comment by lance6716** [comment]
ptal @Ehco1996 @GMHDBJD

**Comment by lance6716** [comment]
I think this change will reduce efforts on other PRs which are related to GTID, so I change the base branch to master. cc @Ehco1996 @lichunzhu @GMHDBJD 

will resolve conflict later.

**Review by Ehco1996 — APPROVED** [review]
LGTM ~ nice work

**Pull request merged**